### PR TITLE
Add permission policy engine

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Add `PermissionEngine` with default tool/path/shell policy evaluation and C-compatible approval callback support.
- Wire permission engine into agent tool execution while preserving existing per-tool approval callbacks as fallback.
- Persist always-allow/deny decisions to `~/.makai/permissions.json` (or configured path) and reload them for future auto-approval.

## Test plan
- [x] `zig build --build-file /Users/lsm/.neokai/projects/makai-c058549f/worktrees/t3-permission-policy-engine/zig/build.zig test-unit-agent-types`
- [x] `zig build --build-file /Users/lsm/.neokai/projects/makai-c058549f/worktrees/t3-permission-policy-engine/zig/build.zig test-unit-agent-loop`
- [x] `zig build --build-file /Users/lsm/.neokai/projects/makai-c058549f/worktrees/t3-permission-policy-engine/zig/build.zig test-unit-agent`
- [x] `zig build --build-file /Users/lsm/.neokai/projects/makai-c058549f/worktrees/t3-permission-policy-engine/zig/build.zig test`

### CODE REVIEW

File: zig/src/tools/permission.zig [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+
+    fn findPersisted(self: *Self, call: ToolCall) ?PermissionDecision {
+        for (self.persisted.items) |decision| {
+            if (matchesPersisted(decision, call)) return decision.decision;
+        }
+        return null;
+    }
+
+    fn isInsideWorkspace(self: *Self, path: []const u8) bool {
+        if (!std.fs.path.isAbsolute(path)) return true;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig]: Fixed. Relative paths now normalize against workspace root and reject traversal that escapes before workspace policy runs; added test for ../../etc/passwd denial.

File: zig/src/tools/permission.zig [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+    if (a == null or b == null) return false;
+    return std.mem.eql(u8, a.?, b.?);
+}
+
+fn pathWithinRoot(path: []const u8, root: []const u8) bool {
+    if (std.mem.eql(u8, path, root)) return true;
+    if (!std.mem.startsWith(u8, path, root)) return false;
+    if (root.len == 0) return false;
+    if (root[root.len - 1] == std.fs.path.sep) return true;
+    return path.len > root.len and path[root.len] == std.fs.path.sep;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig:501]: Fixed. Absolute paths now normalize dot segments before root-prefix comparison; added test for /workspace/../tmp/outside.txt denial.

File: zig/src/agent/agent_loop.zig [hunk 3]
```diff
@@ -527,48 +483,42 @@ fn executeToolCalls(
...
             };
-            if (t.approval_ui_fn) |notify| {
-                notify(t.approval_ui_ctx, approval_request, allocator);
-            }
-            if (t.approval_fn) |approval| {
-                const decision = approval(t.approval_ctx, approval_request);
-                if (decision == .reject) {
+            if (config.permission_engine) |engine| {
+                const decision = try engine.approve(tool_call.name, validated_args);
+                if (decision == .reject or decision == .reject_always) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent_loop.zig]: Fixed. Agent loop now evaluates engine policy first; prompt decisions fall back to existing tool approval callbacks when the engine has no approval callback.

File: zig/src/tools/permission.zig [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+}
+
+fn isSafeShell(command: []const u8) bool {
+    const trimmed = std.mem.trim(u8, command, " \t\r\n");
+    return std.mem.eql(u8, trimmed, "echo") or std.mem.startsWith(u8, trimmed, "echo ");
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig:507]: Fixed. Echo auto-allow now rejects shell control characters, so compound commands like `echo ok; cat /etc/passwd` require prompt; added regression test.

File: zig/src/tools/permission.zig [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+    var command: ?[]u8 = null;
+    errdefer if (command) |owned| allocator.free(owned);
+
+    if (parsed.value == .object) {
+        const obj = parsed.value.object;
+        if (firstStringField(obj, &.{ "path", "file_path", "target_path", "cwd" })) |value| {
+            path = try allocator.dupe(u8, value);
+        }
+        if (firstStringField(obj, &.{ "command", "cmd", "script" })) |value| {
+            command = try allocator.dupe(u8, value);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig:346]: Fixed. Always decisions now persist only when a path/command scope was extracted; scope-less approve_always/reject_always degrade to one-shot approve/reject. Added no-wildcard regression test.

File: zig/src/agent/agent_loop.zig [hunk 6]
```diff
@@ -527,48 +483,42 @@ fn executeToolCalls(
...
+                .args_json = validated_args,
             };
-            if (t.approval_ui_fn) |notify| {
-                notify(t.approval_ui_ctx, approval_request, allocator);
-            }
-            if (t.approval_fn) |approval| {
-                const decision = approval(t.approval_ctx, approval_request);
-                if (decision == .reject) {
+            if (config.permission_engine) |engine| {
+                const decision = try engine.approve(tool_call.name, validated_args);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent_loop.zig]: Fixed. PermissionEngine.approve now treats malformed args as a single-call rejection instead of propagating parse errors; agent loop also uses evaluate before approve. Added malformed JSON regression test.

File: zig/src/tools/permission.zig [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+}
+
+fn isSafeShell(command: []const u8) bool {
+    const trimmed = std.mem.trim(u8, command, " \t\r\n");
+    if (containsShellControl(trimmed)) return false;
+    return std.mem.eql(u8, trimmed, "echo") or std.mem.startsWith(u8, trimmed, "echo ");
+}
+
+fn containsShellControl(command: []const u8) bool {
+    return std.mem.indexOfAny(u8, command, ";&|`$()<>") != null;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig]: Fixed. Newline and carriage-return separators now count as shell control characters, so echo commands with embedded newline-delimited commands fall back to prompt instead of auto-allow. Added regression coverage for `echo hello\ncat /etc/passwd`.

File: zig/src/agent/types.zig [hunk 8]
```diff
@@ -211,10 +212,7 @@ pub const ToolOutputMiddlewareFn = *const fn (
     allocator: std.mem.Allocator,
 ) anyerror!void;
 
-pub const ToolApprovalDecision = enum {
-    approve,
-    reject,
-};
+pub const ToolApprovalDecision = permission.ApprovalDecision;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/types.zig:215]: Fixed. TUI approval wrapper now treats both `.reject` and `.reject_always` from original agent callbacks as blocking decisions, while `.approve`/`.approve_always` continue through. Updated wrapper test to cover `.reject_always`.

File: zig/src/tools/permission.zig [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+
+fn getStringField(obj: std.json.ObjectMap, key: []const u8) ?[]const u8 {
+    const value = obj.get(key) orelse return null;
+    return if (value == .string) value.string else null;
+}
+
+fn matchesPersisted(decision: PersistedDecision, call: ToolCall) bool {
+    if (!std.mem.eql(u8, decision.tool_name, call.tool_name)) return false;
+    if (decision.operation != call.operation) return false;
+    if (!optionalStringEql(decision.path, call.path)) return false;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig:415]: Fixed. Persisted file path scopes now normalize through the same workspace-aware path normalization before storage and before matching. Added coverage proving `a.txt`, `./a.txt`, and `/workspace/a.txt` match the same persisted decision.

File: zig/src/tools/permission.zig [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+            .allow => return .approve,
+            .deny => return .reject,
+            .prompt => {},
+        }
+
+        const callback = self.approval_callback orelse return .reject;
+        const decision = callback(ToolCallC.fromToolCall(call));
+        switch (decision) {
+            .approve_always => if (canPersistDecision(call)) try self.persistDecision(call, .allow) else return .approve,
+            .reject_always => if (canPersistDecision(call)) try self.persistDecision(call, .deny) else return .reject,
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig]: Fixed. `.approve_always`/`.reject_always` persistence failures now degrade to immediate one-shot approve/reject results instead of bubbling I/O errors into the agent loop. Added regression coverage using an unwritable persistence path.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: ✅ APPROVE

All 10 findings from two prior review rounds have been addressed and verified. Tests pass cleanly.

#### Round 1 (6 P1 findings) — All fixed
1. **Relative path traversal** — `isInsideWorkspace` now normalizes relative paths via `normalizeToolPath` + `pathEscapesRoot`. Test: `../../etc/passwd` → deny.
2. **Dot-segment bypass in absolute paths** — Both root and path normalized before prefix check. Test: `/workspace/../tmp/outside.txt` → deny.
3. **Legacy approval fallback** — Agent loop three-branch logic: engine callback → legacy `approval_fn` → no approval. `runLegacyApproval` helper extracted.
4. **Shell control characters** — `containsShellControl` blocks `;&|` etc. Test: `echo hello; cat /etc/passwd` → prompt.
5. **Wildcard persist prevention** — `canPersistDecision` guard prevents scope-less `approve_always` from persisting. Test verifies 0 persisted items.
6. **Malformed args graceful handling** — `evaluate` returns `.prompt`, `approve` returns `.reject` on parse failure. No error propagation to agent loop.

#### Round 2 (2 P1 + 2 P2 findings) — All fixed
7. **Newline bypass** — `containsShellControl` now also blocks `\r\n`. Test: `echo hello\ncat /etc/passwd` → prompt.
8. **Approval variants in TUI** — `approveTool` switch handles all 4 enum values (`.approve`, `.reject`, `.approve_always`, `.reject_always`).
9. **Normalized persisted path scope** — Both `persistDecision` and `findPersisted` normalize paths via `normalizePathScope`. Test: `a.txt` matches `/workspace/a.txt` and `./a.txt`.
10. **Persist failure degradation** — `catch return .approve` / `catch return .reject` instead of `try`. Test creates dir-as-path to force write failure, verifies graceful degrade.

#### Assessment
- Implementation covers all acceptance criteria from the task spec.
- Default policy errs on side of prompting (safe default).
- Path normalization is thorough with `..` traversal, dot-segment, and relative path handling.
- Persistence uses atomic replace via `compat.fs.atomicReplace`.
- Good test coverage with edge cases (malformed JSON, persistence failure, path normalization).
- Clean integration with existing agent loop — optional engine, legacy fallback preserved.

File: zig/src/tools/permission.zig [hunk 11]
```diff
@@ -0,0 +1,688 @@
...
+    const operation = inferOperation(tool_name);
+    var path: ?[]u8 = null;
+    errdefer if (path) |owned| allocator.free(owned);
+    var command: ?[]u8 = null;
+    errdefer if (command) |owned| allocator.free(owned);
+
+    if (parsed.value == .object) {
+        const obj = parsed.value.object;
+        if (firstStringField(obj, &.{ "path", "file_path", "target_path", "cwd" })) |value| {
+            path = try allocator.dupe(u8, value);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig]: Fixed. Relative read/write paths now resolve against a call-supplied `workspace_root` before policy classification, so `{\"workspace_root\":\"/tmp\",\"path\":\"secret.txt\"}` is evaluated as `/tmp/secret.txt` and denied outside the engine workspace. Added regression coverage.

File: zig/src/tools/permission.zig [hunk 12]
```diff
@@ -0,0 +1,688 @@
...
+
+fn deinitParsedToolCall(allocator: std.mem.Allocator, call: ToolCall) void {
+    if (call.path) |path| allocator.free(@constCast(path));
+    if (call.command) |command| allocator.free(@constCast(command));
+}
+
+fn inferOperation(tool_name: []const u8) Operation {
+    if (hasAnyToken(tool_name, &.{ "read", "grep", "glob", "list" })) return .read;
+    if (hasAnyToken(tool_name, &.{ "write", "edit", "delete", "move", "rename" })) return .write;
+    if (hasAnyToken(tool_name, &.{ "shell", "bash", "exec", "run" })) return .shell;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tools/permission.zig:367]: Fixed. Tool operation inference now tokenizes tool names on non-alphanumeric boundaries and matches tokens case-insensitively, so names like `thread_run` no longer match `read` by substring. Added regression coverage for `thread_run` staying prompt while `file_read` remains read/allow.

File: zig/src/agent/agent_loop.zig [hunk 13]
```diff
@@ -550,12 +560,34 @@ fn executeToolCalls(
...
+                if (policy_decision == .prompt and engine.approval_callback != null) {
+                    const decision = try engine.approve(tool_call.name, validated_args);
+                    if (decision == .reject or decision == .reject_always) {
+                        result = try rejectedToolResult(allocator);
+                        is_error = true;
+                        try finalizeToolExecution(allocator, config, event_stream, &results, tool_call, execution_args, &result, is_error);
+                        continue;
+                    }
+                } else if (policy_decision == .prompt) {
+                    const legacy_decision = runLegacyApproval(t, approval_request, allocator);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent_loop.zig]: Fixed. Agent tool execution now runs legacy approval callbacks before permission-engine policy handling, so tool-level callbacks can still reject calls even when the engine returns allow or deny. Added regression coverage for an in-workspace read allowed by policy but rejected by legacy approval.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: ✅ APPROVE

All 13 findings across all review rounds resolved. Commit `80a4cf5` addresses the 3 latest fixes. All tests pass.

#### Round 3 (3 P1 findings) — All fixed
11. **workspace_root in args** — Relative paths now resolve against call-supplied `workspace_root` before policy check. `{"workspace_root":"/tmp","path":"secret.txt"}` correctly denied outside engine workspace. ✅
12. **Token boundary matching** — `inferOperation` now tokenizes on non-alphanumeric boundaries with case-insensitive matching. `thread_run` no longer spuriously matches `read` substring. `file_read` correctly matches read tokens. ✅
13. **Legacy approval precedence** — Tool-level `approval_fn` callbacks now run before permission engine evaluation. Individual tools can enforce stricter policies than the engine. Regression test verifies reject from legacy callback blocks even when engine allows. ✅

No remaining findings.
